### PR TITLE
Sync documents for supported languages only

### LINKED ISSUE #56: Document synchronization actions are triggered for unsupported languages
The language client gets activated when working with files whose language matches `getGrammarScopes`. But once activated, document synchronization actions are triggered for all files, because the editor selector function passed to `DocumentSyncAdapter` disregards `getGrammarScopes`.

### PR DESCRIPTION
**Pull request by janos-ss**
Fixes #56

### CODE REVIEW

File: lib/auto-languageclient.js [hunk 1]
```diff
@@ -171,6 +171,18 @@ export default class AutoLanguageClient {
...
 
+  newEditorSelector(editor: atom$TextEditor, basepath: string): atom$TextEditor => boolean {
+    return editor => this._isFileInProject(editor, basepath) && this._isSupportedLanguage(editor);
+  }
+
+  _isFileInProject(editor: atom$TextEditor, basepath: string): boolean {
+    return (editor.getURI() || '').startsWith(basepath);
+  }
+
+  _isSupportedLanguage(editor: atom$TextEditor): boolean {
```
Comment (damieng) [inline lib/auto-languageclient.js]: We should then call this from `shouldStartForEditor`

File: lib/auto-languageclient.js [hunk 2]
```diff
@@ -171,6 +171,18 @@ export default class AutoLanguageClient {
     }
   }
 
+  newEditorSelector(editor: atom$TextEditor, basepath: string): atom$TextEditor => boolean {
```
Comment (damieng) [inline lib/auto-languageclient.js]: Let's rename this `shouldSyncForEditor` and change the second arg to `projectPath` for consistency.

File: lib/auto-languageclient.js [hunk 3]
```diff
@@ -171,6 +171,18 @@ export default class AutoLanguageClient {
...
+    return editor => this._isFileInProject(editor, basepath) && this._isSupportedLanguage(editor);
+  }
+
+  _isFileInProject(editor: atom$TextEditor, basepath: string): boolean {
```
Comment (damieng) [inline lib/auto-languageclient.js]: Ditto here with the second arg being `projectPath`

**Comment by janos-ss** [comment]
I did the requested changes.

I have a question related to this. Why do we have the `_isFileInProject` condition at all? Why not include all open files? Some of my testers found it unexpected to not see issues in files that they just dragged into the Atom window, and so were not part of the project. What would be the harm in including those?

**Comment by damieng** [comment]
When you start a language server you have to give it a 'root path' which is effectively the top folder where project configuration files live.

Subsequent to that they only expect to receive events relating to files beneath that path.

We do currently support if you have more than one folder correctly starting two projects and ensuring the right server gets the right file events.

We don't yet however support ad-hoc opening of a single file. To do that correctly we'd need to identify it wasn't in an existing path and then traverse up the file system folders until we find what we think would likely be the root then start the server with that.  Deciding what constitutes a project root would be language specific (e.g. `.flow` for Flow, `.sln` or `.csproj` for C# etc)

**Comment by janos-ss** [comment]
Thanks for the explanation.

Btw I don't understand why `continuous-integration/appveyor/pr` fails. I rebased on `master` and all tests pass for me.

**Pull request merged**